File: python/jsoncanon/jsonpatch.py
"""JSON Patch (RFC 6902) and JSON Merge Patch (RFC 7386).

Both produce a transformed value tree that is then canonicalized. Patch ops are
applied in order; on a failed `test` or any structural error the whole patch
fails (raises). Mirrors jsonpatch.nim. Because the result is canonicalized,
intermediate key order is irrelevant to the output.
"""

from __future__ import annotations

from copy import deepcopy

from .jsonpointer import parse_pointer, resolve_tokens, _array_index
from .parser import JValue, JSONError
from .serializer import dumps


def _add(doc: JValue, path: str, value: JValue) -> JValue:
    toks = parse_pointer(path)
    if not toks:
        return value  # add/replace at root
    parent = resolve_tokens(doc, toks[:-1])
    last = toks[-1]
    if isinstance(parent, list):
        i = _array_index(last, len(parent), allow_dash=True)
        if i > len(parent):
            raise JSONError(f"patch add: index {last} out of range")
        parent.insert(i, value)
    elif isinstance(parent, dict):
        parent[last] = value
    else:
        raise JSONError("patch add: parent is not an array or object")
    return doc


def _replace(doc: JValue, path: str, value: JValue) -> JValue:
    toks = parse_pointer(path)
    if not toks:
        return value
    parent = resolve_tokens(doc, toks[:-1])
    last = toks[-1]
    if isinstance(parent, list):
        i = _array_index(last, len(parent))
        if i >= len(parent):
            raise JSONError(f"patch replace: index {last} out of range")
        parent[i] = value
    elif isinstance(parent, dict):
        if last not in parent:
            raise JSONError(f"patch replace: no member {last!r}")
        parent[last] = value
    else:
        raise JSONError("patch replace: parent is not an array or object")
    return doc


def _remove(doc: JValue, path: str) -> tuple[JValue, JValue]:
    toks = parse_pointer(path)
    if not toks:
        raise JSONError("patch: cannot remove the whole document")
    parent = resolve_tokens(doc, toks[:-1])
    last = toks[-1]
    if isinstance(parent, list):
        i = _array_index(last, len(parent))
        if i >= len(parent):
            raise JSONError(f"patch remove: index {last} out of range")
        return parent.pop(i), doc
    if isinstance(parent, dict):
        if last not in parent:
            raise JSONError(f"patch remove: no member {last!r}")
        return parent.pop(last), doc
    raise JSONError("patch remove: parent is not an array or object")


def _op(doc: JValue, op: JValue) -> JValue:
    if not isinstance(op, dict) or "op" not in op or "path" not in op:
        raise JSONError("patch: each operation needs 'op' and 'path'")
    o, path = op["op"], op["path"]
    if not isinstance(o, str) or not isinstance(path, str):
        raise JSONError("patch: 'op' and 'path' must be strings")
    if o == "add":
        return _add(doc, path, deepcopy(op["value"]))
    if o == "replace":
        return _replace(doc, path, deepcopy(op["value"]))
    if o == "remove":
        return _remove(doc, path)[1]
    if o == "move":
        frm = op["from"]
        assert isinstance(frm, str)
        val, doc = _remove(doc, frm)
        return _add(doc, path, val)
    if o == "copy":
        frm = op["from"]
        assert isinstance(frm, str)
        return _add(doc, path, deepcopy(resolve_tokens(doc, parse_pointer(frm))))
    if o == "test":
        got = resolve_tokens(doc, parse_pointer(path))
        if dumps(got) != dumps(op["value"]):
            raise JSONError(f"patch test failed at {path}")
        return doc
    raise JSONError(f"patch: unknown op {o!r}")


def apply_patch(doc: JValue, patch: JValue) -> JValue:
    if not isinstance(patch, list):
        raise JSONError("JSON Patch must be an array of operations")
    doc = deepcopy(doc)
    for op in patch:
        doc = _op(doc, op)
    return doc


def apply_merge_patch(target: JValue, patch: JValue) -> JValue:
    if isinstance(patch, dict):
        base = target if isinstance(target, dict) else {}
        for k, v in patch.items():
            if v is None:
                base.pop(k, None)
            else:
                base[k] = apply_merge_patch(base.get(k), v)
        return base
    return deepcopy(patch)

File: nim/src/jsonpatch.nim
## JSON Patch (RFC 6902) and JSON Merge Patch (RFC 7386). Mirror of
## python/jsoncanon/jsonpatch.py. Produces a transformed value tree that the
## caller then canonicalizes (so intermediate key order is irrelevant).

import ./jsoncanon
import ./jsonpointer

proc deepCopy*(n: JNode): JNode =
  case n.kind
  of jkNull: result = JNode(kind: jkNull)
  of jkBool: result = JNode(kind: jkBool, b: n.b)
  of jkNum: result = JNode(kind: jkNum, num: n.num)
  of jkStr: result = JNode(kind: jkStr, str: n.str)
  of jkNonFinite: result = JNode(kind: jkNonFinite, nf: n.nf)
  of jkArr:
    result = JNode(kind: jkArr)
    for x in n.arr: result.arr.add deepCopy(x)
  of jkObj:
    result = JNode(kind: jkObj)
    for i in 0 ..< n.keys.len:
      result.keys.add n.keys[i]
      result.vals.add deepCopy(n.vals[i])

proc member(o: JNode, key: string): JNode =
  let i = o.keys.find(key)
  if i >= 0: o.vals[i] else: nil

proc opStr(op: JNode, key: string): string =
  let v = member(op, key)
  if v == nil or v.kind != jkStr:
    raise newException(JsonError, "patch: '" & key & "' must be a string")
  v.str

proc parentOf(doc: JNode, path: string): (JNode, string) =
  let toks = parsePointer(path)
  if toks.len == 0: raise newException(JsonError, "patch: empty path has no parent")
  (resolveTokens(doc, toks[0 ..< toks.len - 1]), toks[^1])

proc addOp(doc: JNode, path: string, value: JNode): JNode =
  let toks = parsePointer(path)
  if toks.len == 0: return value
  let (parent, last) = parentOf(doc, path)
  case parent.kind
  of jkArr:
    let i = arrayIndex(last, parent.arr.len, allowDash = true)
    if i > parent.arr.len: raise newException(JsonError, "patch add: index out of range")
    parent.arr.insert(value, i)
  of jkObj:
    let ix = parent.keys.find(last)
    if ix >= 0: parent.vals[ix] = value
    else: (parent.keys.add last; parent.vals.add value)
  else: raise newException(JsonError, "patch add: parent is not an array or object")
  doc

proc replaceOp(doc: JNode, path: string, value: JNode): JNode =
  let toks = parsePointer(path)
  if toks.len == 0: return value
  let (parent, last) = parentOf(doc, path)
  case parent.kind
  of jkArr:
    let i = arrayIndex(last, parent.arr.len)
    if i >= parent.arr.len: raise newException(JsonError, "patch replace: index out of range")
    parent.arr[i] = value
  of jkObj:
    let ix = parent.keys.find(last)
    if ix < 0: raise newException(JsonError, "patch replace: no member '" & last & "'")
    parent.vals[ix] = value
  else: raise newException(JsonError, "patch replace: parent is not an array or object")
  doc

proc removeOp(doc: JNode, path: string): (JNode, JNode) =
  let (parent, last) = parentOf(doc, path)
  case parent.kind
  of jkArr:
    let i = arrayIndex(last, parent.arr.len)
    if i >= parent.arr.len: raise newException(JsonError, "patch remove: index out of range")
    let v = parent.arr[i]
    parent.arr.delete(i)
    (v, doc)
  of jkObj:
    let ix = parent.keys.find(last)
    if ix < 0: raise newException(JsonError, "patch remove: no member '" & last & "'")
    let v = parent.vals[ix]
    parent.keys.delete(ix); parent.vals.delete(ix)
    (v, doc)
  else: raise newException(JsonError, "patch remove: parent is not an array or object")

proc applyOp(doc: JNode, op: JNode): JNode =
  if op.kind != jkObj or member(op, "op") == nil or member(op, "path") == nil:
    raise newException(JsonError, "patch: each operation needs 'op' and 'path'")
  let o = opStr(op, "op")
  let path = opStr(op, "path")
  case o
  of "add": addOp(doc, path, deepCopy(member(op, "value")))
  of "replace": replaceOp(doc, path, deepCopy(member(op, "value")))
  of "remove": removeOp(doc, path)[1]
  of "move":
    let frm = opStr(op, "from")
    let (val, d) = removeOp(doc, frm)
    addOp(d, path, val)
  of "copy":
    let frm = opStr(op, "from")
    addOp(doc, path, deepCopy(resolveTokens(doc, parsePointer(frm))))
  of "test":
    let got = resolveTokens(doc, parsePointer(path))
    if serialize(got) != serialize(member(op, "value")):
      raise newException(JsonError, "patch test failed at " & path)
    doc
  else: raise newException(JsonError, "patch: unknown op '" & o & "'")

proc applyPatch*(doc, patch: JNode): JNode =
  if patch.kind != jkArr:
    raise newException(JsonError, "JSON Patch must be an array of operations")
  result = deepCopy(doc)
  for op in patch.arr: result = applyOp(result, op)

proc applyMergePatch*(target, patch: JNode): JNode =
  if patch.kind == jkObj:
    result = if target.kind == jkObj: target else: JNode(kind: jkObj)
    for i in 0 ..< patch.keys.len:
      let k = patch.keys[i]
      let v = patch.vals[i]
      let existing = result.keys.find(k)
      if v.kind == jkNull:
        if existing >= 0: (result.keys.delete(existing); result.vals.delete(existing))
      else:
        let cur = if existing >= 0: result.vals[existing] else: JNode(kind: jkNull)
        let merged = applyMergePatch(cur, v)
        if existing >= 0: result.vals[existing] = merged
        else: (result.keys.add k; result.vals.add merged)
  else:
    result = deepCopy(patch)